pulp_container/app/access_policy.py

```diff
@@ -32,32 +32,40 @@ class DistributionAccessPolicyMixin:
     Access policy mixin for DistributionViewSet which handles namespace permissions.
     """
 
-    def has_manage_namespace_dist_perms(self, request, view, action, permission):
+    def has_namespace_obj_perms(self, request, view, action, permission):
         """
-        Check whether a user can create a namespace or it can manage distributions
-        for existing namespace.
+        Check if a user has object-level perms on the namespace associated with the distribution
+        or repository.
         """
-        namespace = request.data["base_path"].split("/")[0]
-        return NamespacePermissionsChecker.has_permissions(namespace, request.user, permission)
+        obj = view.get_object()
+        if type(obj) == models.ContainerDistribution:
+            namespace = obj.namespace
+        else:
+            distribution = models.ContainerDistribution.objects.get(repository=obj)
+            namespace = distribution.namespace
+        return request.user.has_perm(permission, namespace)
 
-    def has_namespace_obj_perms(self, request, view, action, permission):
+    def has_namespace_or_obj_perms(self, request, view, action, permission):
         """
-        Check if a user has object-level perms on the namespace associated with the distribution.
+        Check if a user has a namespace-level perms or object-level permission
         """
-        namespace = view.get_object().namespace
-        return request.user.has_perm(permission, namespace)
+        ns_perm = "container.namespace_{}".format(permission.split(".", 1)[1])
+        if self.has_namespace_obj_perms(request, view, action, ns_perm):
+            return True
+        else:
+            return request.user.has_perm(permission, view.get_object())
 
     def obj_exists(self, request, view, action):
         """
         Check if the distribution exists.
         """
         return view.get_object() is not None
 
-    def is_private(self, request, view, action):
+    def is_public(self, request, view, action):
```

Comment:
I think this is very confusing, that you need to set the `private` flag on the distribution to trigger this `is_public` check.